### PR TITLE
SPI: improvements for SPI library

### PR DESCRIPTION
**Pull request by stas2z**
My suggestion for #898

- optimize spi_transfer routine to use LL instead of HAL
- add SPISettings field for transmit only operations
- replace spi_send with call to spi_transfer with SPI_TRANSMITONLY flag
- allow SPI_TRANSFER_TIMEOUT redefinition

### CODE REVIEW

**Comment by ABOSTM** [comment]
Hi @stas2z
I review the PR
That is nice job and looks good to me.
Nevertheless I propose to factorize common code to improve readability and maintenance.
I does not impact performances. (tested on NUCLEO_L476RG with LCD/SD sadafruit shield).
See PR on your own fork https://github.com/stas2z/Arduino_Core_STM32/pull/1

**Comment by uzi18** [comment]
such blind send should be possible with dma, but don't know if any compatibility issues wil rise

**Comment by stas2z** [comment]
> such blind send should be possible with dma, but don't know if any compatibility issues wil rise

it's not a blind send, it checks for the state before DR register access, this is exactly what HAL_SPI_TransmitRecieve does, but shorter and much faster cuz much overhead stuff dropped.

File: libraries/SPI/src/utility/spi_com.c [hunk 1]
```diff
@@ -407,24 +393,34 @@ spi_status_e spi_send(spi_t *obj, uint8_t *Data, uint16_t len, uint32_t Timeout)
   * @param  rx_buffer : data to receive
   * @param  len : length in byte of the data to send and receive
   * @param  Timeout: Timeout duration in tick
+  * @param  skipRecieve: skip recieving data after transmit or not
```
Comment (fpistm) [inline libraries/SPI/src/utility/spi_com.c]:
```suggestion
  * @param  skipReceive: skip receiving data after transmit or not
```

File: libraries/SPI/src/utility/spi_com.c [hunk 2]
```diff
@@ -407,24 +393,34 @@ spi_status_e spi_send(spi_t *obj, uint8_t *Data, uint16_t len, uint32_t Timeout)
...
-
-  hal_status = HAL_SPI_TransmitReceive(&(obj->handle), tx_buffer, rx_buffer, len, Timeout);
-
-  if (hal_status == HAL_TIMEOUT) {
-    ret = SPI_TIMEOUT;
-  } else if (hal_status != HAL_OK) {
-    ret = SPI_ERROR;
+  tickstart = HAL_GetTick();
+  while (size--) {
+    while (!LL_SPI_IsActiveFlag_TXE(_SPI))
```
Comment (fpistm) [inline libraries/SPI/src/utility/spi_com.c]: This LL API is not available in H7 and MP1.

File: libraries/SPI/src/utility/spi_com.c [hunk 3]
```diff
@@ -407,24 +393,34 @@ spi_status_e spi_send(spi_t *obj, uint8_t *Data, uint16_t len, uint32_t Timeout)
...
+      ;
+    LL_SPI_TransmitData8(_SPI, *tx_buffer++);
+
+    if (!skipReceive) {
+      while (!LL_SPI_IsActiveFlag_RXNE(_SPI))
```
Comment (fpistm) [inline libraries/SPI/src/utility/spi_com.c]: This LL API is not available in H7 and MP1.

File: libraries/SPI/src/utility/spi_com.c (hunk 2 above)
Comment (stas2z) [reply libraries/SPI/src/utility/spi_com.c]: ive checked and found *_ll_spi.* is available for H7 and MP1
but yes, i found some differences for H7 at least, will check it

**Comment by stas2z** [comment]
Unfortunately i have no H7 or MP1 board to check

**Comment by ABOSTM** [comment]
Thanks @stas2z I will test H7 and MP1

**Comment by ABOSTM** [comment]
@stas2z,
Tests on MP1 fails.
But here is the fix:
https://github.com/stas2z/Arduino_Core_STM32/pull/2

**Pull request merged**